packages/pwa-buildpack/lib/WebpackTools/plugins/RootComponentsPlugin.js

```diff
@@ -0,0 +1,217 @@
+        debug('apply: subscribing to beforeRun and watchRun');
+        compiler.hooks.beforeRun.tapPromise('RootComponentsPlugin', inject);
+        compiler.hooks.watchRun.tapPromise('RootComponentsPlugin', inject);
+    }
+    async injectRootComponentLoader() {
+        debug('injectRootComponentLoader: running this.buildFetchModule');
+        await this.buildFetchModule();
+        debug('applying VirtualModulePlugin and ProvidePlugin');
+        new VirtualModulePlugin({
+            moduleName: 'FETCH_ROOT_COMPONENT',
+            contents: this.contents
+        }).apply(this.compiler);
+        new ProvidePlugin({
+            fetchRootComponent: 'FETCH_ROOT_COMPONENT'
+        }).apply(this.compiler);
+    }
+    findRootComponentsIn(dir) {
+        return new Promise(resolve => {
+            const jsFiles = [];
+            const done = () => resolve(jsFiles);
+            const fs = this.fs;
+            walk(dir, {
+                fs: {
+                    readdir: fs.readdir.bind(fs),
+                    stat: fs.stat.bind(fs)
+                }
+            })
+                .on('readable', function() {
+                    let item;
+                    while ((item = this.read())) {
+                        if (
+                            item.stats.isFile() &&
+                            extensionRE.test(item.path)
+                        ) {
+                            debug(`will check ${item.path} for RootComponent`);
+                            jsFiles.push(item.path);
+                        }
+                    }
+                })
+                .on('error', done)
+                .on('end', done);
+        });
+    }
+    async buildFetchModule() {
+        const { context, rootComponentsDirs } = this.opts;
+
+        // Create a list of absolute paths for root components. When a
+        // relative path is found, resolve it from the root context of
+        // the webpack build
+        const rootComponentsDirsAbs = rootComponentsDirs.map(dir =>
+            isAbsolute(dir) ? dir : join(context, dir)
+        );
+        debug(
+            'buildFetchModule: absolute root component dirs: %o',
+            rootComponentsDirsAbs
+        );
+        const rootComponentImporters = await rootComponentsDirsAbs.reduce(
+            async (importersPromise, rootComponentDir) => {
+                debug('gathering files from %s', rootComponentDir);
+                const [importerSources, rootComponentFiles] = await Promise.all(
+                    [
+                        importersPromise,
+                        this.findRootComponentsIn(rootComponentDir)
+                    ]
+                );
+                debug(
+                    'files from %s: %o',
+                    rootComponentDir,
+                    rootComponentFiles
+                );
+                await Promise.all(
+                    rootComponentFiles.map(async rootComponentFile => {
```

Comment:
Can probably filter out or ignore `__tests__` files.

`  pwa-buildpack:WebpackTools:plugins:RootComponentsPlugin.js will check /Users/rugh/code/pwa-studio/packages/venia-concept/src/RootComponents/Category/__tests__/categoryContent.spec.js for RootComponent +0ms`

Reply:
Dang, you're right